endaq/device/devinfo.py

```diff
@@ -188,46 +198,75 @@ def _readManifest(self) \
         manFile = os.path.join(self.device.path, self.device._MANIFEST_FILE)
         calFile = os.path.join(self.device.path, self.device._SYSCAL_FILE)
 
-        manData = calData = propData = None
+        self._manData = self._manData = self._manData = None
 
         try:
             with open(manFile, 'rb') as f:
-                manData = f.read()
+                self._manData = f.read()
         except (FileNotFoundError, AttributeError) as err:
             logger.debug(f"Possibly-allowed exception when reading {manFile}: {err!r}")
 
         try:
             with open(calFile, 'rb') as f:
-                calData = f.read()
+                self._calData = f.read()
         except (FileNotFoundError, AttributeError) as err:
             logger.debug(f"Possibly-allowed exception when reading {calFile}: {err!r}")
 
         try:
             # _propData is read and cached here but parsed in `getSensors()`.
             # Old EFM32 recorders stored this w/ the manifest in the USERPAGE.
             with open(self.device.recpropFile, 'rb') as f:
-                propData = f.read()
+                self._propData = f.read()
         except (FileNotFoundError, AttributeError) as err:
             logger.debug("Possibly-allowed exception when reading "
                          f"{self.device.recpropFile}: {err!r}")
 
-        return manData, calData, propData
-
 
-    def readManifest(self) \
-            -> Tuple[Union[None, bytearray, bytes], Union[None, bytearray, bytes], Union[None, bytearray, bytes]]:
+    def readManifest(self) -> Optional[bytes]:
         """ Read the device's manifest data. The data is a superset of the
             information returned by `getInfo()`.
+
+            :return: The raw, EBML encoded manifest data (or `None` if no
+                manifest data are available).
         """
+        if self._manData:
+            return self._manData
+
         if os.path.exists(os.path.join(self.device.path, self.device._USERPAGE_FILE % 0)):
-            return  self._readUserpage()
+            self._readUserpage()
         elif os.path.exists(os.path.join(self.device.path, self.device._MANIFEST_FILE)):
-            return self._readManifest()
+            self._readManifest()
 
-        return None, None, None
+        return self._manData
 
 
-    def readUserCalibration(self) -> Union[None, bytearray, bytes]:
+    def readCalibration(self) -> Optional[bytes]:
+        """ Read the device's factory calibration data.
+
+            :return: The raw, EBML encoded calibration data (or `None` if no
+            calibration data are available).
+        """
+        if self._calData:
+            return self._calData
+
+        self.readManifest()
+        return self._calData
```

Comment:
Has this been tested on a device that uses `manifest` and `syscal` files? It looks like it will only work on a device that uses userpg

Reply:
See `readManifest()` - it calls `_readUserpage()` if there are `USERPG*` files, `_readManifest()` if not. Both methods read and cache the calibration and properties.